hail/src/main/scala/is/hail/expr/types/physical/PCanonicalLocus.scala

```diff
@@ -0,0 +1,103 @@
+package is.hail.expr.types.physical
+import is.hail.variant.ReferenceGenome
+
+import is.hail.annotations._
+import is.hail.asm4s.{Code, coerce}
+import is.hail.backend.BroadcastValue
+import is.hail.expr.ir.EmitMethodBuilder
+import is.hail.expr.types.virtual.TLocus
+import is.hail.utils._
+import is.hail.variant._
+
+
+object PCanonicalLocus {
+  def apply(rg: ReferenceGenome): PLocus = PCanonicalLocus(rg.broadcastRG)
+
+  def apply(rg: ReferenceGenome, required: Boolean): PLocus = PCanonicalLocus(rg.broadcastRG, required)
+
+  def representation(required: Boolean = false): PStruct = PStruct(
+    required,
+    "contig" -> PString(required = true),
+    "position" -> PInt32(required = true))
+
+  def schemaFromRG(rg: Option[ReferenceGenome], required: Boolean = false): PType = rg match {
+    case Some(ref) => PCanonicalLocus(ref)
+    case None => representation(required)
+  }
+}
+
+final case class PCanonicalLocus(rgBc: BroadcastRG, required: Boolean = false) extends PLocus {
+    def rg: ReferenceGenome = rgBc.value
+
+    def _asIdent = "locus"
+    def _toPretty = s"Locus($rg)"
+
+    override def pyString(sb: StringBuilder): Unit = {
+      sb.append("locus<")
+      sb.append(prettyIdentifier(rg.name))
+      sb.append('>')
+    }
+
+    def copy(required: Boolean = this.required) = PCanonicalLocus(this.rgBc, required)
+
+    val representation: PStruct = PCanonicalLocus.representation(required)
+
+    def contig(region: Code[Region], off: Code[Long]): Code[Long] = representation.loadField(region, off, 0)
+
+    def position(region: Code[Region], off: Code[Long]): Code[Int] = Region.loadInt(representation.loadField(region, off, 1))
+
+  // FIXME: Remove when representation of contig/position is a naturally-ordered Long
+  override def unsafeOrdering(): UnsafeOrdering = {
+    val repr = representation.fundamentalType
+
+    val localRGBc = rgBc
+    val binaryOrd = repr.fieldType("contig").asInstanceOf[PBinary].unsafeOrdering()
+
+    new UnsafeOrdering {
+      def compare(r1: Region, o1: Long, r2: Region, o2: Long): Int = {
+        val cOff1 = repr.loadField(r1, o1, 0)
+        val cOff2 = repr.loadField(r2, o2, 0)
+
+        if (binaryOrd.compare(r1, cOff1, r2, cOff2) == 0) {
+          val posOff1 = repr.loadField(r1, o1, 1)
+          val posOff2 = repr.loadField(r2, o2, 1)
+          java.lang.Integer.compare(Region.loadInt(posOff1), Region.loadInt(posOff2))
+        } else {
+          val contig1 = PString.loadString(r1, cOff1)
+          val contig2 = PString.loadString(r2, cOff2)
+          localRGBc.value.compare(contig1, contig2)
+        }
+      }
+    }
+  }
+
+  def codeOrdering(mb: EmitMethodBuilder, other: PType): CodeOrdering = {
+    assert(other isOfType this)
+    new CodeOrderingCompareConsistentWithOthers {
+      type T = Long
+      val contigBin = representation.fundamentalType.fieldType("contig").asInstanceOf[PBinary]
```

Comment:
@tpoterba could you help me understand this cast? We don't have any guarantees that a PString and a PBinary will be compatible types right (though the canonical implementation of both assume this). So I think this should be PCanonicalBinary when that exists, correct?

Reply:
the fundamental type of PString is PBinary, that's why we could do this

Reply:
Duh, I see this now. Fundamental type copies over all of the fields defined on locus, and sets the type in the PField to the fundamental type of the backing type, which for PString is PBinary.